core-java/src/main/java/eu/ostrzyciel/jelly/core/internal/ProtoEncoderImpl.java

```diff
@@ -0,0 +1,115 @@
+package eu.ostrzyciel.jelly.core.internal;
+
+import eu.ostrzyciel.jelly.core.*;
+import eu.ostrzyciel.jelly.core.proto.v1.Rdf;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+public class ProtoEncoderImpl<TNode, TTriple, TQuad> extends ProtoEncoder<TNode, TTriple, TQuad> {
+
+    private boolean hasEmittedOptions = false;
+    private final List<Rdf.RdfStreamRow> rowBuffer;
+
+    protected ProtoEncoderImpl(
+        NodeEncoder<TNode> nodeEncoder,
+        ProtoEncoderConverter<TNode, TTriple, TQuad> converter,
+        ProtoEncoder.Params params
+    ) {
+        super(nodeEncoder, converter, params);
+        this.rowBuffer = Optional.ofNullable(appendableRowBuffer).orElse(new ArrayList<>());
+    }
+
+    @Override
+    public Iterable<Rdf.RdfStreamRow> addTripleStatement(TNode subject, TNode predicate, TNode object) {
```

Comment:
Makes me think – can we make these methods implement the same iface as the one we'd be calling in ProtoDecoder?

See the core-patch module. I have there the PatchHandler iface. It's extended by the encoder, and used by the decoder. Clean, shiny. Can we do the same here?

Reply:
doned